File: packages/core/src/research_engine/domain/edges.py
"""Edge/relationship domain types."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from research_engine.domain.common import NodeKind


class Edge(BaseModel):
    """A directed edge between two nodes."""

    id: UUID
    source_kind: NodeKind
    source_id: UUID
    target_kind: NodeKind
    target_id: UUID
    relation_type: str
    attributes: dict[str, Any] = Field(default_factory=dict)
    source_passage_id: UUID | None = None
    confidence: float = 1.0
    created_at: datetime


class EdgeDraft(BaseModel):
    """Data needed to create an edge."""

    source_kind: NodeKind
    source_id: UUID
    target_kind: NodeKind
    target_id: UUID
    relation_type: str
    attributes: dict[str, Any] = Field(default_factory=dict)
    source_passage_id: UUID | None = None
    confidence: float = 1.0

File: crates/marginalia-types/src/edges.rs
//! Edge/relationship types mirroring `domain/edges.py`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::common::NodeKind;

/// A directed edge between two nodes.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Edge {
    pub id: Uuid,
    pub source_kind: NodeKind,
    pub source_id: Uuid,
    pub target_kind: NodeKind,
    pub target_id: Uuid,
    pub relation_type: String,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    #[serde(default)]
    pub source_passage_id: Option<Uuid>,
    #[serde(default = "default_confidence")]
    pub confidence: f64,
    pub created_at: DateTime<Utc>,
}

fn default_confidence() -> f64 {
    1.0
}

/// Data needed to create an edge.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EdgeDraft {
    pub source_kind: NodeKind,
    pub source_id: Uuid,
    pub target_kind: NodeKind,
    pub target_id: Uuid,
    pub relation_type: String,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    #[serde(default)]
    pub source_passage_id: Option<Uuid>,
    #[serde(default = "default_confidence")]
    pub confidence: f64,
}
